lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -292,22 +295,26 @@ def compile(node, options)
     config
   end
 
-  # Turn our host name into a node object.
-  def find_node(name, environment, transaction_uuid, configured_environment)
+  # Use indirection to find the node associated with a given request
+  def find_node(facts, request)
+    name = request.key || request.node
+    environment = request.environment
+    transaction_uuid = request.options[:transaction_uuid]
+    configured_environment = request.options[:configured_environment]
+
     Puppet::Util::Profiler.profile(_("Found node information"), [:compiler, :find_node]) do
       node = nil
       begin
         node = Puppet::Node.indirection.find(name, :environment => environment,
                                              :transaction_uuid => transaction_uuid,
-                                             :configured_environment => configured_environment)
+                                             :configured_environment => configured_environment,
+                                             :facts => facts)
```

Comment:
Should we update the various built-in node terminii to use this new parameter? At a glance, `plain` and `ldap` at least will query the facts terminus (using `Node::fact_merge`)